File: 算法/QuickSort.js
/**
 * 非递归实现快速排序
 * @param {*} data 
 * @param {*} low 
 * @param {*} high 
 */
function quickSort(data, low, high) {
    console.log(data, low, high)
    const pivot = partition(data, low, high);
    let result = [];

    if (pivot + 1 < high) {
        result.push(pivot + 1, high);
    } else if (pivot > low + 1) {
        result.push(low, pivot - 1);
    }



    // 数组不为空
    while (result.length > 0) {
        console.log(1, result)
        const temp_high = result.pop();
        const temp_low = result.pop();
        if (temp_low >= temp_high) {
            break;
        }
        const temp_pivot = partition(data, temp_low, temp_high);

        if (temp_pivot + 1 < high) {
            result.push(temp_pivot + 1, temp_high)
        } else if (temp_pivot - 1 > low) {
            result.push(temp_low, temp_pivot - 1);
        }

    }

    return data;

}


// 严版获取快排pivot
function partition(data, low, high) {

    let pivot = data[low];

    while (low < high) {
        // 高位
        while (low < high && pivot <= data[high]) --high;
        data[low] = data[high]
        // 低位
        while (low < high && pivot >= data[low]) ++low;
        data[high] = data[low]
    }
    data[low] = pivot

    return low
}

console.log(quickSort([1, 8, 9, 2], 0, 3))

File: code/ds/c++版本/QuickSort.cpp
// 快速排序【伪代码】
void QuickSort(ElemType A[] , int low , int high){
  // low > high 表角标越界，low=high 子表只有一个元素，不需要进行快排，已经有序
  if(low<high){

    // 获取pivot基准，将当前待排序表分成左右两个子表
    int pivotKey = Partition(A,low,high)

    // 对左边序列进行快排
    QuickSort(A,low,pivotKey-1)

    // 对右边序列进行快排
    QuickSort(A,pivotKey+1,high)

  }
  return A
}

int Partition(ElemType A ,int low , int high){

  ElemType pivot=A[low];

  while(low<high){
    while(low<high && A[high]>=pivot) --high
    A[low]=A[high]  // 比pivot小的都移到左表 注意--high 从后往前遍历

    while(low<high && A[low]<=pivot ) ++low
    A[high]=A[low]  // 比pivot大的都移到右表，注意++low 从前往后遍历
  }

  // 此时low==high||low>high 跳出循环后即找到能将当前表一分为二的pivotKey值
  A[low]=pivot
  // 基准元素pivot对应最终的位置角标
  return low
}